### ISSUE TITLE
Bug: [@typescript-eslint/no-unsafe-return] False positive on unwrapping a generic

### ISSUE THREAD

**Issue by luixo** [typescript-eslint/typescript-eslint#12103]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=C4TwDgpgBA6gTgQzJOAeAKgPigXigbygEsA7EiOALinSgF8BuAKFEigFEAPYRAY2AD2aACJQI3CCQAmAZ1iJkFVAhIhM2HEyhRR44JNnykKVKQBmFKADVsAfmtRq5AG4VmvASRnAoZkrihUXQlpOXhjJRU1TAAKLV8BAWouHgR%2BISDMJgBKXGwY3JxsM0SGIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolrRE%2BWNCbooiaNA7RI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Wrapper<T> = { inner: T };
type Extractor<D extends Wrapper<any>> =
  D extends Wrapper<infer V> ? V : never;
const fn = <D extends Wrapper<any>>(
  foo: Extractor<D>
) => () => foo;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-return": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Resulting `foo` value is of type `Extractor<D>` so I expect rule not to complain

### Actual Result

Rule fails

**Comment by JoshuaKGoldberg** [comment]
👍 this might be a duplicate, or at least related to, #2432. Unclear. Investigation needed. But either way this is a bug that should be fixed, `Extractor<D>` is definitely not `any`. Thanks for filing!
